### PR TITLE
The select_and_scatter GPU kernel was causing a large performance dro…

### PR DESCRIPTION
**Pull request by rsanthanam-amd**
…p for the MLPerf Resnet50 model on MI200.

The reason for the performance loss is that the AMD GPU compiler was inserting very high latency instructions into
the critical path of the CAS loop which was generated because atomic add is not supported for fp16.

These high latency instructions are specific to MI200 are emitted by the compiler to ensure system wide cache
coherency on the MI200 architecture.

The fix for the issue is to change the syncscope for the atomicCAS loop from 'system' (which is the default) to
'agent'.

This prevents the compiler from emitting the MI200 specific high latency instructions but still emits instructions which provide
adequate cache coherency guarantees in line with previous AMD GPU architectures.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 1]
```diff
@@ -310,26 +312,32 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
...
     if (is_atomic_integral) {
       // integral + integral
-      AtomicRMW(llvm::AtomicRMWInst::Add, output_address, source,
-                llvm::MaybeAlign(),
-                llvm::AtomicOrdering::SequentiallyConsistent);
+      if (isGfx90a) {
+        AtomicRMW(llvm::AtomicRMWInst::Add, output_address, source,
+            llvm::MaybeAlign(),
+            llvm::AtomicOrdering::SequentiallyConsistent,
+            b_.getContext().getOrInsertSyncScopeID("agent"));
```
Comment (deven-amd) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:333]: think the "sync-scope" should be set to "agent" irrespective of the gfx type (i.e. set it for all, not just gfx90a)

@whchung @ekuznetsov139 what do you think?

Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc:333]: yeah, i was being cautious.  will wait to hear from others on this.

**Comment by rsanthanam-amd** [comment]
Removed gfx90a restriction on agent sync scope changes.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 2]
```diff
@@ -289,7 +291,7 @@ bool IrEmitter::MaybeEmitDirectAtomicOperation(
       }
     }
 
-    if (target_triple.isAMDGPU()) {
+    if (isAMDGPU) {
       std::string arch = ir_emitter_context_->amdgpu_arch();
       bool isGfx908Plus = arch.size()>=6 &&
```
Comment (whchung) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: `isGfx908Plus` is a dead code?

**Review by whchung — CHANGES_REQUESTED** [review]
sorry for belated reply.
Other than the potential dead code `isGfx908Plus`, the PR looks fine.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc (hunk 2 above)
Comment (rsanthanam-amd) [reply tensorflow/compiler/xla/service/gpu/ir_emitter.cc]: the dead code has been deleted.

**Comment by rsanthanam-amd** [comment]
the 2 commits have been squashed into 1 and the code review comments from @whchung have been addressed.

File: tensorflow/compiler/xla/service/gpu/ir_emitter.cc [hunk 3]
```diff
@@ -507,10 +525,21 @@ Status IrEmitter::EmitAtomicOperationUsingCAS(const HloComputation& computation,
   // Emit code to perform the atomicCAS operation
   // (cas_old_output, success) = atomicCAS(memory_address, cas_old_output,
   //                                       cas_new_output);
-  llvm::Value* ret_value = AtomicCmpXchg(
-      atomic_memory_address, cas_old_output, cas_new_output, llvm::MaybeAlign(),
-      llvm::AtomicOrdering::SequentiallyConsistent,
-      llvm::AtomicOrdering::SequentiallyConsistent);
+  llvm::Value* ret_value = [&]() {
```
Comment (rsanthanam-amd) [inline tensorflow/compiler/xla/service/gpu/ir_emitter.cc:528]: note: this was the change that fixed the original issue.  the other (similar) changes were done to preemptively avoid future performance issues.

**Comment by rsanthanam-amd** [comment]
jenkins: retest Ubuntu-GPU-multi please

**Comment by rsanthanam-amd** [comment]
jenkins: retest Ubuntu-GPU-multi please

**Pull request merged**